[Pipelining] Fix FSDP+PP stream sync bug

wconstab opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #144596
* #144352
* __->__ #144535
* #144534

This bug could cause gradient corruption as a race condition exists
between FSDP's reduce-scatter and any operations reading .grad on the
main stream.  The root cause is that pipelining stage .backward implementation
got modified to support zero-bubble and in doing so, invoked .grad()
instead of .backward(), and performed manual gradient accumulation and
manually called into hooks for FSDP.  But one key hook was missed for
FSDP, the '_root_post_backward_final_callback' hook, which is
responsible for syncing the grad reduction ops after the last layer's
backward completes.

Note: this fix applies to both zero-bubble and non-zero-bubble schedules.  This caused some confusion initially, as non-zero-bubble schedules do use torch.autograd.backward() which would have called into fsdp's hooks and synced, unlike zero-bubble which uses .grad() which does not invoke hooks.  However, this difference was already taken into consideration as FSDP's hooks are manually disabled before invoking either type of backward, and then the hooks are manually triggered.

A better fix as a follow up PR would be to invoke .backward() for the
weight grad, so that we never have to disable or manually invoke hooks.

Modified test_pp_dp to intentionally race against FSDP's reduce by
modifying the parameters inplace in a mathematically identical way, and
confirmed it fails intermittently when the FSDP sync is not applied and
passes with the FSDP sync added.

cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @d4l3k @c-p-i-o

wconstab committed:
[Pipelining] Fix FSDP+PP stream sync bug

This bug could cause gradient corruption as a race condition exists
between FSDP's reduce-scatter and any operations reading .grad on the
main stream.  The root cause is that pipelining stage .backward implementation
got modified to support zero-bubble and in doing so, invoked .grad()
instead of .backward(), and performed manual gradient accumulation and
manually called into hooks for FSDP.  But one key hook was missed for
FSDP, the '_root_post_backward_final_callback' hook, which is
responsible for syncing the grad reduction ops after the last layer's
backward completes.

A better fix as a follow up PR would be to invoke .backward() for the
weight grad, if possible.

Modified test_pp_dp to intentionally race against FSDP's reduce by
modifying the parameters inplace in a mathematically identical way, and
confirmed it fails intermittently when the FSDP sync is not applied and
passes with the FSDP sync added.

[ghstack-poisoned]

awgu commented on torch/distributed/pipelining/stage.py:
```diff
@@ -649,6 +649,11 @@ def run_post_backward(fsdp_module: FSDPModule) -> None:
                         if state._fsdp_param_group:
                             state._fsdp_param_group.post_backward()
 
+                    # it would be much better if pipelining backward invoked .bacward so autograd hooks
```
nit (typo):
```suggestion
                    # it would be much better if pipelining backward invoked .backward so autograd hooks
```

awgu reviewed (approved):
looks good to unblock current numerics

wconstab committed:
Update on "[Pipelining] Fix FSDP+PP stream sync bug"


This bug could cause gradient corruption as a race condition exists
between FSDP's reduce-scatter and any operations reading .grad on the
main stream.  The root cause is that pipelining stage .backward implementation
got modified to support zero-bubble and in doing so, invoked .grad()
instead of .backward(), and performed manual gradient accumulation and
manually called into hooks for FSDP.  But one key hook was missed for
FSDP, the '_root_post_backward_final_callback' hook, which is
responsible for syncing the grad reduction ops after the last layer's
backward completes.

A better fix as a follow up PR would be to invoke .backward() for the
weight grad, if possible.

Modified test_pp_dp to intentionally race against FSDP's reduce by
modifying the parameters inplace in a mathematically identical way, and
confirmed it fails intermittently when the FSDP sync is not applied and
passes with the FSDP sync added.

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 d4l3k c-p-i-o

[ghstack-poisoned]

fegin commented on torch/distributed/pipelining/stage.py:
```diff
@@ -649,6 +649,11 @@ def run_post_backward(fsdp_module: FSDPModule) -> None:
+                    # it would be much better if pipelining backward invoked .backward so autograd hooks
+                    # worked and modules like DDP/FSDP behaved as expected.  Working around this for the time being,
+                    # we need to call this too to ensure FSDP syncs its grad reduction ops back to the default stream.
+                    fsdp_state._root_post_backward_final_callback()
```
What if `perform_backward` calls `stage_backward`? Do we get double hook in such a case?

wconstab replied:
it appears we do. However, according to @awgu it is idempotent so the second call should not be a problem.

Still, this all points to the design being wrong.  We should find a way to make PP invoke .backward() consistently so we don't have to deal with hooks/accumulations manually.

cc @H-Huang

wconstab replied:
sorry, i was confused, but there is no double-hook issue because we _always_ disable the FSDP hooks before calling into perform_backward.  So this PR should be correct, and it should fix correctness for both Zero-bubble and non-zero-bubble schedules since we _always_ manually call FSDP hooks.
<img width="1236" alt="Screenshot 2025-01-10 at 3 15 55 PM" src="https://github.com/user-attachments/assets/18f1171f-1d73-42f6-b783-5cfb54399d28" />

wconstab commented:
@pytorchbot merge
